src/aspire/orientation/__init__.py

```diff
@@ -7,35 +15,434 @@ class Orient3D:
     """
     Define a base class for estimating 3D orientations
     """
-    def __init__(self, src):
+    def __init__(self, src, nrad=None, ntheta=None, masked=True):
         """
-        constructor of an object for estimating 3D orientations
+        Initialize an object for estimating 3D orientations
+
+        :param src: The source object of 2D denoised or class-averaged imag
+        :param nrad: The number of points in the radial direction
+        :param ntheta: The number of points in the theta direction
+        :param masked: Whether applying a mask to the images or not
         """
-        pass
+        self.src = src
+        self.nimg = src.n
+        self.nres = self.src.L
+        if nrad is None:
+            self.nrad = math.ceil(config.orient.r_ratio * self.nres)
+        else:
+            self.nrad = nrad
+        if ntheta is None:
+            self.ntheta = config.orient.n_theta
+        else:
+            self.ntheta = ntheta
+
+        self.basis = PolarBasis2D((self.nres, self.nres), self.nrad, self.ntheta)
+
+        self.max_shift = math.ceil(config.orient.max_shift * self.nres)
+        self.shift_step = config.orient.shift_step
+
+        imgs = self.src.images(start=0, num=np.inf).asnumpy()
 
-    def get_commonlines(self):
+        if masked:
+            mask_radius = self.nres * 0.45
+            self.rise_time = config.orient.rise_time
+            self.fuzzy_mask_dims = config.orient.fuzzy_mask_dims
+            # mask_radius is of the form xxx.5
+            if mask_radius * 2 == int(mask_radius * 2):
+                mask_radius = math.ceil(mask_radius)
+            # mask is not of the form xxx.5
+            else:
+                mask_radius = int(round(mask_radius))
+            # mask projections
+            m = self._fuzzy_mask(self.nres, mask_radius)
+            masked_projs = imgs.copy()
+            masked_projs = masked_projs.transpose((2, 0, 1))
+            masked_projs *= m
+            masked_projs = masked_projs.transpose((1, 2, 0))
+        else:
+            masked_projs = imgs
+
+        # Obtain coefficients in polar Fourier basis for input 2D images
+        self.basis = PolarBasis2D((self.nres, self.nres), self.nrad, self.ntheta)
+        self.pf = self.basis.evaluate_t(masked_projs)
+        self.pf = m_reshape(self.pf, (self.nrad, self.ntheta, self.nimg))
+
+        n_theta = self.ntheta
+        if n_theta % 2 == 1:
+            raise ValueError('n_theta must be even')
+        n_theta = n_theta // 2
+        self.pf = np.concatenate((np.flip(self.pf[1:, n_theta:], 0), self.pf[:, :n_theta]), 0)
```

Comment:
They are not the same thing. From the `basis.evaluate_t` the first two dimensions of `pf` is arranged by  `n_rad` x `n_theta.` In this part of code, it tries to convert pf into an array of size `(2xn_rad-1)` x `n_theta/2,` that is, takes entire ray through the origin, but take the angles only up PI. But in the real calculation of `build_clmatrix,` only half of them are used and speed the computation by a factor of two. More comments added in commit 44ceae2.

Reply:
The statement here is used to rearrange the representation of Fourier transform of images to fit the internal need of other functions such as  `build_clmatrix` and `estimate_shifts`. Of cause we can use the same form from the Polar2DBasis class. Then we need to go through each line in other functions to make corresponding changes. This rearrangement is easier to do.

Reply:
Yes but those other functions (`build_clmatrix` and `estimate_shifts`) only use half of the rays, as you point out, so why do we double it here? It's confusing (and wasteful) to double it only to halve it later on.

Reply:
The output from Polar2DBasis `evalute_t` function have the full representation. We are not double it (n_theta is only half of original value). It only rearrange it in a better way to access in the  `build_clmatrix` and `estimate_shifts`  function to use half of them (there are some kind of differences on how to use). For better memory usage it might be possible to output only half of them in `evaluate_t` and rewrite related code in  `build_clmatrix` and `estimate_shifts`  functions. We also need to make sure other algorithms only use half of them too. This will take quit a lot of time. It will be better to address it on another PR.

Reply:
I'm not talking about having `evaluate_t` outputting half of the *angles*, I'm saying that we're doubling the *radial frequencies* here (with the flip and concatenate), but we only use one half later on – all instances of `pf` are reduced by taking “half” of the radial frequencies (which means we go back to what we had). For example, see here: https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/7b420f9be74de98fb57f7b0a9a83c00aeee2884e/src/aspire/orientation/__init__.py#L582

Reply:
Simply using half of ray doesn't work because not all functions only use half of ray such as applying filters and normalization. In commit  6e2a6d0, all changes were made to use half of ray in all functions.